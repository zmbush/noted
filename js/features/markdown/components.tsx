// Copyright 2021 Zachary Bush.
//
// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
// option. This file may not be copied, modified, or distributed
// except according to those terms.
//

/* eslint-disable react/jsx-props-no-spreading */
import * as React from 'react';
import type { HeadingProps } from 'react-markdown/lib/ast-to-react';
import { Link as RouterLink } from 'react-router-dom';

import {
  Table,
  TableBody,
  TableCell,
  TableContainer,
  TableHead,
  TableRow,
  Typography,
  Link as MuiLink,
} from '@mui/material';

import Directive from './Directive';

export const MaybeDirective = ({ children, node: _node, ...props }: any) => {
  if ('data-directive-type' in props && 'data-type' in props) {
    return (
      <Directive
        // eslint-disable-next-line react/prop-types
        type={(props as any)['data-type']!}
      >
        {children}
      </Directive>
    );
  }
  return <div {...props}>{children}</div>;
};

export const QuoteDirective = ({ children }: any) => (
  <Directive type='blockquote'>{children}</Directive>
);

export const TableRoot = ({ children, node: _node, ...props }: any) => (
  <TableContainer>
    <Table {...props}>{children}</Table>
  </TableContainer>
);

export const THead = ({ children, node: _node, ...props }: any) => (
  <TableHead {...props}>{children}</TableHead>
);

export const TBody = ({ children, node: _node, ...props }: any) => (
  <TableBody
    sx={{
      '@media print': {
        '& tr:nth-of-type(2n + 1)': {
          backgroundColor: '#e0e5c1',
        },
      },
    }}
    {...props}
  >
    {children}
  </TableBody>
);

export const TR = ({ children, node: _node, isHeader: _isHeader, ...props }: any) => (
  <TableRow {...props}>{children}</TableRow>
);

export const TC = ({ children, node: _node, isHeader: _isHeader, ...props }: any) => (
  <TableCell
    sx={{ '@media print': { color: 'black', borderBottom: 'none', padding: 1 } }}
    {...props}
  >
    {children}
  </TableCell>
);

export const Paragraph = ({ children, node: _node, ...props }: any) => (
  <Typography component='p' variant='body' {...props}>
    {children}
  </Typography>
);

export const Heading = ({
  children,
  node: _node,
  level,
  ref: _ref,
  ...props
}: Omit<HeadingProps, 'level'> & { level: 1 | 2 | 3 | 4 }) => {
  const l = (level + 1) as 2 | 3 | 4 | 5;
  return (
    <Typography variant={`h${l}`} {...props}>
      {children}
    </Typography>
  );
};

export const Anchor = ({ children, node: _node, href, ...props }: any) => (
  <MuiLink underline='hover' color='secondary' component={RouterLink} to={href} {...props}>
    {children}
  </MuiLink>
);
