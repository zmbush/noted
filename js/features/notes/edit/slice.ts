// Copyright 2021 Zachary Bush.
//
// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
// option. This file may not be copied, modified, or distributed
// except according to those terms.
//

/* eslint-disable no-param-reassign */
import { createSlice, PayloadAction } from '@reduxjs/toolkit';

export interface NoteEditState {
  editingNote: { id: 'new'; title?: string; parent_note_id?: number } | { id: number } | null;
}

const initialState: NoteEditState = {
  editingNote: null,
};

export const prefix = 'noteEdit';
const noteEditSlice = createSlice({
  name: prefix,
  initialState,
  reducers: {
    setEditingNote(state, { payload }: PayloadAction<NoteEditState['editingNote']>) {
      state.editingNote = payload;
    },
  },
});

export const { setEditingNote } = noteEditSlice.actions;
export default noteEditSlice.reducer;
