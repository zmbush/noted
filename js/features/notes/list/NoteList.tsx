// Copyright 2019 Zachary Bush.
//
// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
// option. This file may not be copied, modified, or distributed
// except according to those terms.
//
import * as React from 'react';
import { useSelector } from 'react-redux';

import { Grid, styled } from '@mui/material';

import { AppState } from 'features/redux/store';

import { NoteWithTags } from 'data/types';

import Note from '../view/Note';

import { getNoteEntities, getSortedNoteIds, getSubNotes } from './selectors';

type Props = {
  depth: number;
  renderOnly?: Set<number>;
  noteViewFilter: { [id: number]: boolean } | null;
  parent_note_id: number | null;
  width?: false | 'auto' | true | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 | 11 | 12;
};

const GridItem = styled(Grid)({
  paddingTop: 0,
  '@media print': {
    padding: '0 !important',
  },
});

const NoteList = ({
  depth,
  renderOnly,
  width = 12,
  noteViewFilter,
  parent_note_id: parentNoteId,
}: Props) => {
  const notesIn = useSelector<AppState, { [id: string]: NoteWithTags }>((state) =>
    renderOnly ? getNoteEntities(state) : getSubNotes(state, { noteId: parentNoteId }),
  );
  const sortedIds = useSelector(getSortedNoteIds);

  let notes = notesIn;
  if (renderOnly) {
    notes = Object.fromEntries(
      Object.entries(notes).filter(([id, _note]) => renderOnly.has(parseInt(id, 10))),
    );
  }

  return (
    <>
      {sortedIds.map((id) => {
        if (id in notes) {
          const n = notes[id];
          if (!noteViewFilter || noteViewFilter[id]) {
            return (
              <GridItem item key={n.id} xs={width}>
                <Note note={n}>
                  <NoteList
                    parent_note_id={n.id}
                    noteViewFilter={noteViewFilter}
                    depth={depth + 1}
                  />
                </Note>
              </GridItem>
            );
          }
        }
        return null;
      })}
    </>
  );
};

export default NoteList;
