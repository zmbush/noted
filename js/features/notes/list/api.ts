// Copyright 2021 Zachary Bush.
//
// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
// option. This file may not be copied, modified, or distributed
// except according to those terms.
//
import { createAsyncThunk } from '@reduxjs/toolkit';

import api from 'data/api';
import { NewNotePayload, NoteWithTags, UpdateNotePayload } from 'data/types';

export const prefix = 'notes';
const name = (n: string) => `${prefix}/${n}`;

export const getNotes = createAsyncThunk(name('list'), async (_: void, { rejectWithValue }) => {
  try {
    return await api.note.list();
  } catch (e) {
    throw rejectWithValue(e);
  }
});

export const updateNote = createAsyncThunk(
  name('update'),
  async (
    {
      id: noteId,
      tags,
      ...note
    }: UpdateNotePayload & Partial<Pick<NoteWithTags, 'tags'>> & { id: number },
    { rejectWithValue },
  ) => {
    try {
      let result = await api.note.update(noteId, note);
      if (tags) {
        result = await api.note.setTags(result.id, tags);
      }
      return result;
    } catch (e) {
      throw rejectWithValue(e);
    }
  },
);
export const createNote = createAsyncThunk(
  name('create'),
  async ({ tags, ...note }: NewNotePayload & Pick<NoteWithTags, 'tags'>, { rejectWithValue }) => {
    try {
      const result = await api.note.create(note);
      return await api.note.setTags(result.id, tags);
    } catch (e) {
      throw rejectWithValue(e);
    }
  },
);

export const deleteNote = createAsyncThunk(
  name('delete'),
  async (noteId: number, { rejectWithValue }) => {
    try {
      await api.note.delete(noteId);
      return noteId;
    } catch (e) {
      throw rejectWithValue(e);
    }
  },
);
