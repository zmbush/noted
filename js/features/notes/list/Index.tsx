// Copyright 2021 Zachary Bush.
//
// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
// option. This file may not be copied, modified, or distributed
// except according to those terms.
//
import * as React from 'react';
import { useSelector } from 'react-redux';

import NoteList from './NoteList';
import { getIsNotArchived } from './selectors';

const Index = () => {
  const noteViewFilter = useSelector(getIsNotArchived);
  return <NoteList noteViewFilter={noteViewFilter} parent_note_id={null} depth={1} />;
};

export default Index;
