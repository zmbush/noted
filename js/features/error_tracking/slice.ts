// Copyright 2021 Zachary Bush.
//
// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
// option. This file may not be copied, modified, or distributed
// except according to those terms.
//

/* eslint-disable no-param-reassign */
import { AsyncThunk, createSlice, PayloadAction } from '@reduxjs/toolkit';

import { prefix as notePrefix } from 'features/notes/list/api';
import { prefix as userPrefix } from 'features/user/api';

import { ErrorData } from 'data/types';

export interface ErrorState {
  any: ErrorData | null;
  [notePrefix]?: ErrorData;
  [userPrefix]?: ErrorData;
}

const initialState: ErrorState = { any: null };

type GenericAsyncThunk = AsyncThunk<unknown, unknown, any>;
type RejectedAction = ReturnType<GenericAsyncThunk['rejected']>;

export const prefix = 'errorTracking';

export const errorTrackingSlice = createSlice({
  name: prefix,
  initialState,
  reducers: {
    clearLastError(state, { payload = 'any' }: PayloadAction<keyof ErrorState | undefined>) {
      if (payload === 'any') {
        state.any = null;
      } else if (payload) {
        delete state[payload];
      }
    },
  },
  extraReducers: (builder) => {
    // Tracking lastError
    builder.addMatcher(
      (action): action is RejectedAction => action?.type?.endsWith('/rejected'),
      (state, action) => {
        const [slice, ..._] = action.type.split('/', 2);
        state.any = action.payload as ErrorData;
        if (slice === userPrefix || slice === notePrefix) {
          state[slice] = action.payload as ErrorData;
        }
      },
    );
  },
});

export const { clearLastError } = errorTrackingSlice.actions;
export default errorTrackingSlice.reducer;
