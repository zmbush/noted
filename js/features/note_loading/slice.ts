// Copyright 2021 Zachary Bush.
//
// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
// option. This file may not be copied, modified, or distributed
// except according to those terms.
//

/* eslint-disable no-param-reassign */
import { createSlice, Draft, PayloadAction } from '@reduxjs/toolkit';

import { updateNote, createNote } from 'features/notes/list/api';

export type NoteLoadingState = { [note_id: number]: string[] };
const initialState: NoteLoadingState = {};

export const prefix = 'noteLoading';

export const uiSlice = createSlice({
  name: prefix,
  initialState,
  reducers: {},
  extraReducers: (builder) => {
    const noteStartEditing = <
      V,
      Ty extends string,
      T extends { id: number } | { parent_note_id?: number | null },
    >(
      state: Draft<NoteLoadingState>,
      action: PayloadAction<V, Ty, { arg: T; requestId: string }>,
    ) => {
      let id: number;
      if ('id' in action.meta.arg) {
        id = action.meta.arg.id;
      } else if ('parent_note_id' in action.meta.arg && action.meta.arg.parent_note_id) {
        id = action.meta.arg.parent_note_id;
      } else {
        // We can't track it.
        return;
      }
      state[id] = [...new Set([...(state[id] || []), action.meta.requestId])];
    };

    const noteDoneEditing = <
      V,
      Ty extends string,
      T extends { id: number } | { parent_note_id?: number | null },
    >(
      state: Draft<NoteLoadingState>,
      action: PayloadAction<V, Ty, { arg: T; requestId: string }>,
    ) => {
      let id: number;
      if ('id' in action.meta.arg) {
        id = action.meta.arg.id;
      } else if ('parent_note_id' in action.meta.arg && action.meta.arg.parent_note_id) {
        id = action.meta.arg.parent_note_id;
      } else {
        // We can't track it.
        return;
      }
      const changing = new Set(state[id] || []);
      changing.delete(action.meta.requestId);
      if (changing.size > 0) {
        state[id] = [...changing];
      } else {
        delete state[id];
      }
    };
    builder.addCase(updateNote.pending, noteStartEditing);
    builder.addCase(createNote.pending, noteStartEditing);

    builder.addCase(updateNote.fulfilled, noteDoneEditing);
    builder.addCase(updateNote.rejected, noteDoneEditing);
    builder.addCase(createNote.fulfilled, noteDoneEditing);
    builder.addCase(createNote.rejected, noteDoneEditing);
  },
});

export default uiSlice.reducer;
