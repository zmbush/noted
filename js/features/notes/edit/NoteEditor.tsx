// Copyright 2019 Zachary Bush.
//
// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
// option. This file may not be copied, modified, or distributed
// except according to those terms.
//
import { createSelector } from '@reduxjs/toolkit';

import * as React from 'react';
import { useDispatch, useSelector } from 'react-redux';

import { Save as SaveIcon } from '@mui/icons-material';
import {
  Card,
  CardContent,
  CardHeader,
  IconButton,
  Input,
  Dialog,
  useMediaQuery,
  Theme,
} from '@mui/material';

import { AppState } from 'features/redux/store';

import BindKeyboard from 'components/BindKeyboard';
import ChipInput from 'components/ChipInput';
import ConfirmationDialog from 'components/ConfirmationDialog';
import Loading from 'components/Loading';
import { NoteWithTags, NewNotePayload } from 'data/types';

import { createNote, updateNote } from '../list/api';
import { getSingleNote } from '../list/selectors';

import { NoteEditState, prefix, setEditingNote } from './slice';

const MarkdownEditor = React.lazy(() => import('features/markdown/Editor'));

type NewNoteOrEdit = { id: 'new'; note: NewNotePayload } | { id: number; note: NoteWithTags };

const getNoteEdit = (state: AppState): NoteEditState => state[prefix];
const getEditingNote = createSelector(getNoteEdit, (noteEdit) => noteEdit.editingNote);
const getNoteBeingEdited = createSelector(
  (s: AppState) => s,
  getEditingNote,
  (state, noteEdit): NewNoteOrEdit | null => {
    if (!noteEdit) {
      return null;
    }
    if (noteEdit.id === 'new') {
      const { id, title = '', parent_note_id: parentNoteId } = noteEdit;
      const note: NewNotePayload = { title, body: '', parent_note_id: parentNoteId };
      return { id, note };
    }
    const { id } = noteEdit;
    return { id, note: getSingleNote(state, { noteId: id }) };
  },
);

const NoteEditorInner = ({ id, note }: NewNoteOrEdit) => {
  const [title, setTitle] = React.useState(note.title);
  const [body, setBody] = React.useState(note.body);
  let noteTags: string[] = [];
  if ('tags' in note) {
    noteTags = note.tags;
  }
  const [tags, setTags] = React.useState(noteTags);
  const [confirmCancelEditOpen, setConfirmCancelEditOpen] = React.useState(false);
  const editorFullscreen = useMediaQuery<Theme>((theme) => theme.breakpoints.down('sm'));
  const dispatch = useDispatch();

  const current = new Map(Object.entries({ title, body, tags }));
  const initial = new Map(Object.entries(note));
  const noteIsModified =
    Array.from(current.keys()).filter((key: string) => current.get(key) !== initial.get(key))
      .length > 0;

  const stopEditing = () => dispatch(setEditingNote(null));
  const save = async (e: React.SyntheticEvent | Event) => {
    e.preventDefault();
    stopEditing();

    if (id === 'new') {
      await dispatch(createNote({ title, body, tags, parent_note_id: note.parent_note_id }));
    } else {
      await dispatch(updateNote({ id, title, body, tags }));
    }
  };

  const tryCancelEdit = () => {
    if (noteIsModified) {
      setConfirmCancelEditOpen(true);
    } else {
      stopEditing();
    }
  };

  return (
    <>
      <ConfirmationDialog
        open={confirmCancelEditOpen}
        title='If you close this editor, you will lose your changes.'
        onPositive={stopEditing}
        onNegative={() => setConfirmCancelEditOpen(false)}
      />
      <Dialog
        open
        fullWidth
        maxWidth='lg'
        fullScreen={editorFullscreen}
        onClose={tryCancelEdit}
        data-testid='edit-note-dialog'
      >
        <BindKeyboard keys='ctrl+s' callback={save}>
          <Card
            data-testid='NoteEditor'
            sx={(theme) => ({
              [theme.breakpoints.up('sm')]: {
                height: '84vh',
              },
              [theme.breakpoints.down('md')]: {
                height: '100vh',
              },
            })}
          >
            <CardHeader
              title={
                <Input
                  sx={{ width: '100%' }}
                  value={title}
                  onChange={(e) => setTitle(e.target.value)}
                  style={{ fontSize: 'inherit' }}
                />
              }
              action={
                <IconButton onClick={save} aria-label='Save Note' size='large' tabIndex={-1}>
                  <SaveIcon />
                </IconButton>
              }
            />
            <CardContent
              sx={{
                position: 'absolute',
                left: 0,
                right: 0,
                bottom: 0,
                top: '50px',
              }}
            >
              <ChipInput
                value={tags}
                onChange={setTags}
                options={['type:Location', 'type:Character']}
              />
              <React.Suspense fallback={<Loading />}>
                <MarkdownEditor
                  height='calc(100% - 50px)'
                  body={body}
                  onChange={(v) => setBody(v())}
                />
              </React.Suspense>
            </CardContent>
          </Card>
        </BindKeyboard>
      </Dialog>
    </>
  );
};

const NoteEditor = () => {
  const noteAndId = useSelector(getNoteBeingEdited);
  if (noteAndId === null) {
    return null;
  }
  const { id } = noteAndId;
  if (id === 'new') {
    const { note } = noteAndId;
    return <NoteEditorInner id='new' note={note} />;
  }
  const { note } = noteAndId;
  return <NoteEditorInner id={id} note={note} />;
};

export default NoteEditor;
