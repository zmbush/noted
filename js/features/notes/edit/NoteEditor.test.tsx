// Copyright 2019 Zachary Bush.
//
// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
// option. This file may not be copied, modified, or distributed
// except according to those terms.
//

/* eslint-disable react/jsx-props-no-spreading */
import userEvent from '@testing-library/user-event';

import React from 'react';

import { signUpUser } from 'features/user/api';

import { render, waitFor } from 'components/test-utils';

import { createNote } from '../list/api';

import NoteEditor from './NoteEditor';
import { setEditingNote } from './slice';

const testNote: Parameters<typeof createNote>[0] = {
  title: 'note title',
  body: `# Note Title
    
This is more content!

## Subheading
`,
  tags: ['tag1'],
};

describe('<NoteEditor />', () => {
  let rendered: ReturnType<typeof render>;
  beforeEach(async () => {
    rendered = render(<NoteEditor />);
    const { store } = rendered;
    await store.dispatch(signUpUser({ email: 'test2@test.com', name: '', password: 'pass' }));
  });

  describe('while editing note', () => {
    beforeEach(async () => {
      const { store } = rendered;
      await store.dispatch(createNote(testNote));
      const { notes } = store.getState();
      expect(notes.ids).toHaveLength(1);
      store.dispatch(setEditingNote({ id: notes.ids[0] as number }));
    });

    test('matches snapshot', async () => {
      const { container, store } = rendered;
      expect(store.getState().errorTracking.any).toBeNull();
      expect(store.getState().notes).toMatchInlineSnapshot(`
        Object {
          "entities": Object {
            "1": Object {
              "archived": false,
              "body": "# Note Title
            
        This is more content!

        ## Subheading
        ",
              "created_at": "",
              "id": 1,
              "parent_note_id": 0,
              "pinned": false,
              "tags": Array [
                "tag1",
              ],
              "title": "note title",
              "updated_at": "",
              "user_id": 2,
            },
          },
          "ids": Array [
            1,
          ],
        }
      `);
      expect(container).toMatchSnapshot();
    });

    test('tag input works', async () => {
      const { getByTestId, getByText, queryByText, store } = rendered;
      const tagsInput = getByText('tag1');
      userEvent.type(tagsInput, 'A Tag{enter}Another Tag{enter}A Third Tag{enter}');
      expect(getByText('A Tag')).toMatchInlineSnapshot(`
              <span
                class="MuiChip-label MuiChip-labelMedium css-6od3lo-MuiChip-label"
              >
                A Tag
              </span>
          `);
      expect(getByText('Another Tag')).toMatchInlineSnapshot(`
              <span
                class="MuiChip-label MuiChip-labelMedium css-6od3lo-MuiChip-label"
              >
                Another Tag
              </span>
          `);
      expect(getByText('A Third Tag')).toMatchInlineSnapshot(`
              <span
                class="MuiChip-label MuiChip-labelMedium css-6od3lo-MuiChip-label"
              >
                A Third Tag
              </span>
          `);
      // Click delete button
      userEvent.click(getByText('A Third Tag').parentElement!.children[1]);
      expect(queryByText('A Third Tag')).toBeNull();

      getByText('A Tag');
      userEvent.click(getByText('A Tag').parentElement!.children[1]);
      expect(queryByText('A Tag')).toBeNull();

      userEvent.click(getByTestId('SaveIcon'));
      waitFor(() =>
        expect(store.getState().notes.entities[store.getState().notes.ids[0]]!.tags).toEqual([
          'tag1',
          'Another Tag',
        ]),
      );
    });

    test('title change works', () => {
      const { getByDisplayValue, getByTestId, store } = rendered;
      const titleInput = getByDisplayValue(
        store.getState().notes.entities[store.getState().notes.ids[0]]!.title,
      );

      expect(titleInput).toMatchInlineSnapshot(`
      <input
        class="MuiInput-input MuiInputBase-input css-1x51dt5-MuiInputBase-input-MuiInput-input"
        type="text"
        value="note title"
      />
    `);
      userEvent.type(titleInput, ' More Title');
      expect(titleInput).toMatchInlineSnapshot(`
      <input
        class="MuiInput-input MuiInputBase-input css-1x51dt5-MuiInputBase-input-MuiInput-input"
        type="text"
        value="note title More Title"
      />
    `);

      userEvent.click(getByTestId('SaveIcon'));
      waitFor(() =>
        expect(store.getState().notes.entities[store.getState().notes.ids[0]]!.title).toEqual(
          'note title More Title',
        ),
      );
    });

    test('renders body as expected', () => {
      const { getAllByRole } = render(<NoteEditor />);
      const bodyInput = getAllByRole('textbox')[1];

      expect(bodyInput).toMatchInlineSnapshot(`
      <div
        class="ProseMirror"
        contenteditable="true"
        role="textbox"
      >
        <a
          class="heading-name ProseMirror-widget"
          id="h-note-title"
        />
        <h1>
          <span
            class="heading-actions "
            contenteditable="false"
          >
            <button
              class="heading-anchor"
              type="button"
            />
            <button
              class="heading-fold "
              type="button"
            >
              <svg
                fill="currentColor"
                height="24"
                viewBox="6 0 12 24"
                width="12"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M8.23823905,10.6097108 L11.207376,14.4695888 L11.207376,14.4695888 C11.54411,14.907343 12.1719566,14.989236 12.6097108,14.652502 C12.6783439,14.5997073 12.7398293,14.538222 12.792624,14.4695888 L15.761761,10.6097108 L15.761761,10.6097108 C16.0984949,10.1719566 16.0166019,9.54410997 15.5788477,9.20737601 C15.4040391,9.07290785 15.1896811,9 14.969137,9 L9.03086304,9 L9.03086304,9 C8.47857829,9 8.03086304,9.44771525 8.03086304,10 C8.03086304,10.2205442 8.10377089,10.4349022 8.23823905,10.6097108 Z"
                />
              </svg>
            </button>
          </span>
          <span
            class="heading-content"
          >
            Note Title
          </span>
        </h1>
        <p>
          This is more content!
        </p>
        <a
          class="heading-name ProseMirror-widget"
          id="h-subheading"
        />
        <h2>
          <span
            class="heading-actions "
            contenteditable="false"
          >
            <button
              class="heading-anchor"
              type="button"
            />
            <button
              class="heading-fold "
              type="button"
            >
              <svg
                fill="currentColor"
                height="24"
                viewBox="6 0 12 24"
                width="12"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M8.23823905,10.6097108 L11.207376,14.4695888 L11.207376,14.4695888 C11.54411,14.907343 12.1719566,14.989236 12.6097108,14.652502 C12.6783439,14.5997073 12.7398293,14.538222 12.792624,14.4695888 L15.761761,10.6097108 L15.761761,10.6097108 C16.0984949,10.1719566 16.0166019,9.54410997 15.5788477,9.20737601 C15.4040391,9.07290785 15.1896811,9 14.969137,9 L9.03086304,9 L9.03086304,9 C8.47857829,9 8.03086304,9.44771525 8.03086304,10 C8.03086304,10.2205442 8.10377089,10.4349022 8.23823905,10.6097108 Z"
                />
              </svg>
            </button>
          </span>
          <span
            class="heading-content"
          >
            Subheading
          </span>
        </h2>
      </div>
    `);
    });
  });
});
