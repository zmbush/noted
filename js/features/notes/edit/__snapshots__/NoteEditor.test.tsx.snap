// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NoteEditor /> while editing note matches snapshot 2`] = `
<div
  aria-hidden="true"
/>
`;
